src/parse.rs

```diff
@@ -37,18 +41,174 @@ impl Expr {
     /// rnix::SyntaxNode isn't recognized, we don't get tooling for its children.
     pub fn parse(node: SyntaxNode, scope: Gc<Scope>) -> Result<Self, EvalError> {
         let range = Some(node.text_range());
-        let recurse = |node| Expr::parse(node, scope.clone()).map(|x| Box::new(x));
+        let recurse_box = |node| Expr::parse(node, scope.clone()).map(|x| Box::new(x));
+        let recurse_gc = |node| Expr::parse(node, scope.clone()).map(|x| Gc::new(x));
         let source = match ParsedType::try_from(node.clone()).map_err(|_| ERR_PARSING)? {
+            ParsedType::Select(select) => ExprSource::Select {
+                from: recurse_gc(select.set().ok_or(ERR_PARSING)?),
+                index: recurse_box(select.index().ok_or(ERR_PARSING)?),
+            },
+            ParsedType::AttrSet(set) => {
+                let is_recursive = set.recursive();
+
+                // Create a new scope if we're a recursive attr set. We'll later
+                // populate this scope with the non-dynamic keys of the set.
+                let new_scope = if is_recursive {
+                    let new = Scope::Normal {
+                        parent: scope.clone(),
+                        contents: GcCell::new(HashMap::new()),
+                    };
+                    Gc::new(new)
+                } else {
+                    scope.clone()
+                };
+
+                let mut map = HashMap::new();
+                let mut definitions = vec![];
+                let mut inherits = vec![];
+
+                // Construct key-value pairs. We store the original path syntax so the user
+                // can hover over it to inspect dynamic values in code like:
+                // `{ foo."${toString (1+1)}".bar = 2; }`.
+                for entry in set.entries() {
+                    let inner = Self::parse(entry.value().ok_or(ERR_PARSING)?, new_scope.clone())?;
+
+                    let path = entry
+                        .key()
+                        .ok_or(ERR_PARSING)?
+                        .path()
+                        .map(|node| Self::parse(node, scope.clone()).map(Box::new))
+                        .collect::<Vec<_>>();
+
+                    let root_key = path[0].as_ref()?.as_ident()?;
+
+                    // Although we need to preserve the original path syntax for user friendliness,
+                    // we'll still convert the value to nested maps, for the sake of simplicity.
+                    let value = if path.len() == 1 {
+                        inner
+                    } else {
+                        let mut out = inner;
+                        for part in path.iter().skip(1).rev() {
+                            let mut map = HashMap::new();
```

Comment:
Naming a variable `map` if there's another one in the outer scope seems a bit confusing.